components/service-operator/internal/aws/cloudformation/client.go

```diff
@@ -0,0 +1,477 @@
+			StackName: aws.String(stack.GetStackName()),
+		})
+		if err != nil {
+			return err
+		}
+	}
+	_, err = r.waitUntilDestroyedState(ctx, stack)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// Outputs fetches the cloudformation outputs for the given stack
+// Returns ErrStackNotFound if stack does not exist
+func (r *Client) Outputs(ctx context.Context, stack Stack) (Outputs, error) {
+	state, err := r.get(ctx, stack)
+	if err != nil {
+		return nil, err
+	}
+	return r.resolveOutputs(ctx, state.Outputs)
+}
+
+// resolveOutputs returns cloudformation outputs in a map format and resolves
+// any values that are stored in AWS Secrets Manager
+func (r *Client) resolveOutputs(ctx context.Context, list []*Output) (Outputs, error) {
+	outputs := Outputs{}
+	for _, item := range list {
+		if item.OutputKey == nil || item.OutputValue == nil {
+			continue
+		}
+		key := *item.OutputKey
+		value := *item.OutputValue
+		// we automatically resolve references to AWS Secrets Manager
+		// secrets here, so that we are able to make use of encrypted
+		// sensitive values in cloudformation templates
+		if strings.HasPrefix(value, "{{resolve:secretsmanager:") {
+			// extract ARN and key name from reference
+			secretARNMatcher := regexp.MustCompile(`{{resolve:secretsmanager:(.*):SecretString:(.*)}}`)
+			matches := secretARNMatcher.FindStringSubmatch(value)
+			if len(matches) == 0 {
+				return nil, fmt.Errorf("failed to extract ARN and key name from secretsmanager value: %s", secretARNMatcher)
+			}
+			arn := matches[1]
+			subkey := matches[2]
+			v, err := r.Client.GetSecretValueWithContext(ctx, &GetSecretValueInput{
+				SecretId: aws.String(arn),
+			})
+			if err != nil {
+				return nil, err
+			}
+			if v.SecretString == nil {
+				return nil, fmt.Errorf("unexpected nil value in SecretString of %s", arn)
+			}
+			secrets := map[string]interface{}{}
+			err = json.Unmarshal([]byte(*v.SecretString), &secrets)
+			if err != nil {
+				return nil, err
+			}
+			subval, haveSubkey := secrets[subkey]
+			if !haveSubkey {
+				return nil, fmt.Errorf("could not find subkey %s in SecretString of %s", subkey, arn)
+			}
+			subvalString, ok := subval.(string)
+			if !ok {
+				return nil, fmt.Errorf("subval at subkey %s in SecretString of %s is not a string", subkey, arn)
+			}
+			value = subvalString
+		}
+		outputs[key] = value
+	}
+	return outputs, nil
+}
+
+// get fetches the cloudformation stack state
+// Returns ErrStackNotFound if stack does not exist
+func (r *Client) get(ctx context.Context, stack Stack) (*State, error) {
+	describeOutput, err := r.Client.DescribeStacksWithContext(ctx, &DescribeStacksInput{
+		StackName: aws.String(stack.GetStackName()),
+	})
+	if err != nil {
+		if IsNotFoundError(err) {
+			return nil, ErrStackNotFound
+		}
+		return nil, err
+	}
+	if describeOutput == nil {
+		return nil, fmt.Errorf("describeOutput was nil, potential issue with AWS Client")
+	}
+	if len(describeOutput.Stacks) == 0 {
```

Comment:
these three errors are catching things that the AWS SDK documentation say shouldn't happen (and mainly exist to avoid a `panics` that arrose when creating the fake aws mock). checking for >1 seems like a sensible sanity check, I'll add that.

Reply:
done